### ISSUE TITLE
AVIATOR

### ISSUE THREAD

**Issue by Wilen047** [Wilen047/AVIATOR-PREDICTION-APK#11]
# Issue #11: Develop Machine Learning Model for Historical Data Analysis

**Status:** Open

**Priority:** P1

**Labels:** `machine-learning`, `model`, `training`, `scikit-learn`, `analytics`

## Description

Develop and train machine-learning models using the feature dataset prepared in Issue #10.

The purpose of this issue is to evaluate whether historical features contain useful statistical information for modeling future multiplier values. The model must not be presented as a guaranteed Aviator prediction system because future game outcomes are intended to be random.

The model should be evaluated using proper time-based validation and compared against simple baseline methods.

## Objectives

* Define the ML problem.
* Select appropriate target variables.
* Prepare training and testing datasets.
* Establish baseline models.
* Train candidate ML models.
* Compare model performance.
* Prevent data leakage.
* Save the best validated model.
* Track model versions and metadata.

## ML Problem Definition

Start with two possible approaches.

### Approach A: Regression

Predict a numerical multiplier-related target.

Example:

```text
Input:
Historical features

Output:
Estimated future multiplier
```

The result must be described as a **model estimate**, not a guaranteed outcome.

### Approach B: Classification

Classify the next observation into configurable multiplier ranges.

Example:

```text
Class 1 → 1.00x – 1.49x
Class 2 → 1.50x – 1.99x
Class 3 → 2.00x – 2.99x
Class 4 → 3.00x+
```

Classification should only be used if the dataset and evaluation results justify it.

## Target Definition

Define the target carefully so that it represents a future observation.

For example:

```text
target = next_round_multiplier
```

or:

```text
target_class = next_round_multiplier_range
```

The target must be shifted so that features from the current time do not contain information from the future target.

## Dataset Splitting

Do not use random train/test splitting for the main time-series experiment.

Use chronological splitting:

```text
Historical Data
      ↓
Training Data
      ↓
Validation Data
      ↓
Test Data
```

Example:

```text
70% → Training
15% → Validation
15% → Testing
```

The exact percentages can be adjusted depending on dataset size.

## Baseline Models

Before using complex ML models, establish simple baselines.

Examples:

### Regression Baselines

* Mean predictor.
* Median predictor.
* Previous-value baseline.

### Classification Baselines

* Most-frequent class.
* Simple threshold-based classifier.

The ML model must demonstrate whether it actually improves on these baselines.

## Candidate Models

Experiment with suitable models such as:

### Regression

* Linear Regression.
* Random Forest Regressor.
* Gradient Boosting Regressor.

### Classification

* Logistic Regression.
* Random Forest Classifier.
* Gradient Boosting Classifier.

Only retain complex models if they provide meaningful improvement during validation.

## Model Training Pipeline

The training workflow should be:

```text
Feature Dataset
      ↓
Target Generation
      ↓
Chronological Split
      ↓
Training
      ↓
Validation
      ↓
Hyperparameter Selection
      ↓
Final Training
      ↓
Test Evaluation
      ↓
Model Selection
      ↓
Model Versioning
```

### Regression

Calculate:

* MAE
* RMSE
* R²

Example:

```text
MAE  = 0.XX
RMSE = X.XX
R²   = 0.XX
```

### Classification

Calculate:

* Accuracy.
* Precision.
* Recall.
* F1-score.
* Confusion matrix.
* Log loss where appropriate.

Because class distributions may be imbalanced, accuracy should not be the only metric.

## Time-Series Validation

Use time-aware validation where appropriate.

For example:

```text
Fold 1:
Train → Earlier data
Validate → Later data

Fold 2:
Train → Larger historical period
Validate → Subsequent period
```

This better represents how the model would operate on unseen future observations.

## Data Leakage Prevention

The training process must ensure:

* Future observations are not used as features.
* Test data is not used during model selection.
* Scaling is fitted only on training data.
* Feature selection does not use test labels.
* Hyperparameter tuning does not use the final test set.

## Model Comparison

Create a comparison table internally containing:

```text
Model
Metric
Training Score
Validation Score
Test Score
Training Time
Version
```

Example:

```text
Random Forest
MAE: 0.XX
RMSE: X.XX

Gradient Boosting
MAE: 0.XX
RMSE: X.XX
```

The best model should be selected based on predefined evaluation criteria rather than simply choosing the most complex model.

## Model Versioning

Each trained model must have a unique version.

Example:

```text
model_name: AviatorHistoricalModel
model_version: 1.0.0
algorithm: RandomForestRegressor
```

Store metadata such as:

```text
model_version
algorithm
features_used
training_dataset
training_date
validation_metrics
test_metrics
```

## Model Storage

Save trained models using an appropriate serialization method.

Example:

```text
models/
├── aviator_model_v1.pkl
└── metadata_v1.json
```

Model files should not contain sensitive user information.

## Implementation Structure

Suggested structure:

```text
app/
├── ml/
│   ├── models/
│   │   ├── baseline.py
│   │   ├── regression.py
│   │   └── classification.py
│   │
│   ├── training/
│   │   ├── train.py
│   │   ├── evaluate.py
│   │   └── validation.py
│   │
│   └── model_registry.py
│
└── schemas/
    └── model.py
```

## Training Endpoint

Model training should preferably run as a background/admin operation rather than during normal user requests.

Possible endpoint:

```text
POST /api/v1/ml/train
```

Only administrators should be allowed to trigger training.

## Testing Requirements

* [ ] Target generation works.
* [ ] Chronological splitting works.
* [ ] Baseline models are implemented.
* [ ] Candidate models train successfully.
* [ ] Regression metrics are calculated.
* [ ] Classification metrics are calculated where applicable.
* [ ] Time-aware validation works.
* [ ] Data leakage checks pass.
* [ ] Model comparison works.
* [ ] Model versioning works.
* [ ] Best model can be saved.
* [ ] Saved model can be loaded successfully.

## Acceptance Criteria

* [ ] ML problem is clearly defined.
* [ ] Target variable is correctly generated.
* [ ] Training/validation/test datasets are chronological.
* [ ] Baseline model is implemented.
* [ ] At least two candidate models are evaluated where appropriate.
* [ ] Evaluation metrics are implemented.
* [ ] Time-aware validation is implemented.
* [ ] Data leakage prevention is verified.
* [ ] Models are compared objectively.
* [ ] Best validated model is saved.
* [ ] Model metadata is stored.
* [ ] Model version is recorded.
* [ ] Training pipeline is reproducible.
* [ ] Tests pass.

## Expected Outcome

A reproducible machine-learning training pipeline that determines whether historical data provides measurable predictive signal and produces a properly evaluated model-based estimate system.

## Important Note

If evaluation shows that the model performs no better than simple baselines or fails to generalize to unseen chronological data, the system should report that result honestly rather than presenting unreliable predictions as accurate.

## Next Issue

**Issue #12: Model Evaluation, Validation & Performance Analysis**
